include/ppx/knob.h

```diff
@@ -0,0 +1,377 @@
+
+    KnobSlider(const std::string& flagName, T defaultValue, T minValue, T maxValue)
+        : Knob(flagName)
+    {
+        PPX_ASSERT_MSG(minValue < maxValue, "invalid range to initialize slider");
+        PPX_ASSERT_MSG(minValue <= defaultValue && defaultValue <= maxValue, "defaultValue is out of range");
+
+        mMinValue = minValue;
+        mMaxValue = maxValue;
+        SetDefaultAndValue(defaultValue);
+    }
+
+    void Draw() override
+    {
+        ImGui::SliderInt(mDisplayName.c_str(), &mValue, mMinValue, mMaxValue, NULL, ImGuiSliderFlags_AlwaysClamp);
+    }
+
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    std::string GetFlagHelpText() const override
+    {
+        std::string flagHelpText = "--" + mFlagName + " <" + std::to_string(mMinValue) + "~" + std::to_string(mMaxValue) + ">";
+        if (!mFlagDesc.empty()) {
+            flagHelpText += " : " + mFlagDesc;
+        }
+        return flagHelpText + "\n";
+    }
+
+    // Expected commandline flag format:
+    // --flag_name <int>
+    void UpdateFromFlags(const CliOptions& opts) override
+    {
+        SetDefaultAndValue(opts.GetExtraOptionValueOrDefault(mFlagName, mValue));
+    }
+
+    T GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    void SetValue(T newVal)
+    {
+        if (!IsValidValue(newVal)) {
+            PPX_LOG_ERROR(mFlagName << " cannot be set to " << newVal << " because it's out of range " << mMinValue << "~" << mMaxValue);
+            return;
+        }
+
+        mValue = newVal;
+    }
+
+private:
+    bool IsValidValue(T val)
+    {
+        if (val < mMinValue || val > mMaxValue)
+            return false;
+        return true;
+    }
+
+    void SetDefaultAndValue(T newVal)
+    {
+        PPX_ASSERT_MSG(IsValidValue(newVal), "invalid default value");
+        mDefaultValue = newVal;
+        ResetToDefault();
+    }
+
+private:
+    T mValue;
+    T mDefaultValue;
+
+    // mValue will be clamped to the mMinValue to mMaxValue range, inclusive
+    T mMinValue;
+    T mMaxValue;
+};
+
+// KnobDropdown will be displayed as a dropdown in the UI
+// The knob stores the index of a selected choice from a list of allowed options
+template <typename T>
+class KnobDropdown final
+    : public Knob
+{
+public:
+    static_assert(std::is_same_v<T, std::string>, "KnobDropdown must be created with type: std::string");
+
+    template <std::input_iterator Iter>
+    KnobDropdown(
+        const std::string& flagName,
+        size_t             defaultIndex,
+        Iter               choicesBegin,
+        Iter               choicesEnd)
+        : Knob(flagName)
+    {
+        PPX_ASSERT_MSG(static_cast<int>(defaultIndex) >= 0, "defaultIndex must be non-negative");
+        PPX_ASSERT_MSG(static_cast<int>(defaultIndex) < std::distance(choicesBegin, choicesEnd), "defaultIndex is out of range");
```

Comment:
It only create problem if you actually give it an `input_iterator` (e.g. terminal input) instead of `forward_iterator`.

Also please don't cast `defaultIndex` to `int`, `mChoices.size()` is `std::vector<T>::size_type` (almost always just `size_t` in a different name) and is unsigned. Same as other comment, please note that `int` might be 32 bit on some 64 bit systems.